### DISCUSSION TITLE
리팩터링 2판 서평

### DISCUSSION THREAD

**Discussion by choisohyun** [frontend-book-study/refactoring-2#19] [리팩터링 2판]
```
## 📑 서평
### 한 줄 요약
[이 책의 핵심 메시지나 가장 큰 인상을 한 문장으로]

### 주요 내용 정리
[책의 핵심 내용과 인상 깊었던 부분을 정리해주세요]

### 인상 깊은 구절
> [인용구를 적어주세요]

나의 생각: [이 구절에 대한 나의 해석과 느낌]

### 비평적 관점
**좋았던 점**
- 
- 

**아쉬웠던 점**
- 

## ⭐ 종합적 평가
### 책에 대한 평가
- 전체 평점: [5점 만점]
- 추천하고 싶은 대상:
- 이유:

```

**Comment by RAINBOWJYP** [comment]
## 📑 서평
### 한 줄 요약
좋은 리팩터링의 습관 중 하나는 테스트다.

### 주요 내용 정리
- 리팩터링을 자주 하도록 노력해야 한다.
- 악취를 구분하려면 부록 B를 보자
- 컴파일 - 테스트 - 커밋 > 컴파일 - 테스트 - 커밋 > 컴파일 - 테스트 - 커밋 >...
- 주석이 필요 없는 코드로 리팩터링 해보기
- 반복문을 파이프라인으로 바꿔서 시대에 걸맞은 반복문을 만들기
- 나중에 필요할거야!라는 생각으로 작성한 코드는 눈앞에서 치워버리기


### 인상 깊은 구절
> 하나의 값이 여러 목적으로 사용된다면 혼란과 버그를 낳는다.

나의 생각: React.useState가 생각이 났다. 하나의 setState를 몇 가지 경우에 활용하는 경우가 있었는데 조심해야겠다는 생각을 했다.


### 비평적 관점
**좋았던 점**
- 리팩터링은 한번에 확 바꾸기보다는 점진적으로 한스텝씩 진행해야함을 보여준다. 이 과정에서 테스트는 필수이다. 
- 시작부터 리팩터링 기법을 하나씩 알려주기보단 모든 기법을 활용하여 리팩터링하는 예시를 먼저 보여주고 각 기법에 대해 설명하는 흐름이 좋았다.
- 리팩터링의 중요성을 깨닫게되고 책을 보면서 내 코드에서 어떻게 리팩터링 하면 좋을지 아이디어가 떠오르기도 해서 좋았다.

**아쉬웠던 점**
- 수많은 리팩터링 기법들을 소개해주는데 내가 작성된 코드를 이해해야(클래스를 알아야) 어떤 리팩터링 기법이 적합할지 알 수 있을 것 같다.  
- 예시를 자세히 적으면 책 두께가 두꺼워져서 그랬던걸까? 예시를 이해하려면 전과 후를 왔다갔다하며 이해를 해야했다.
- 리팩터링 기법 번역이 아쉽다. 영어로 써져있는 기법명을 보고 아~ 하게된 것들이 있다.

## ⭐ 종합적 평가
### 책에 대한 평가
- 전체 평점: ⭐⭐⭐⭐ (이해못한거 많아서 별하나뺌)
- 추천하고 싶은 대상: 중니어 이상의 개발자
- 이유: 주니어인 나는 클래스 문법에 익숙하지 않아 절반을 이해하지 못했다. 클래스에 익숙한 분들이라면 재밌게 읽을 수 있을 것 같다!

**Comment by LastCow9000** [comment]
## 📑 서평
### 한 줄 요약
함수명, 변수명을 잘 짓고 책임에 맞게 함수, 객체, 클래스로 분리하자!

### 주요 내용 정리
- 컴파일 - 테스트 - 커밋
- 리팩터링을 하면 기존 코드를 읽기 쉽고 수정하기 쉽게 되기 때문에 개발 속도가 오히려 빨라진다.
- 리팩터링은 비슷한 일을 세 번째 하게 되면 한다(삼진 리팩터링)
- 기술을 잘 모르는 관리자에게는 "리팩터링한다고 말하지 말라"
- 한 줄의 코드라도 함수로 빼내서 무엇을 하는 지 명확하게 표현할 수 있다면 추출하는 것이 좋다.
- 컬렉션 변수로의 접근을 캡슐화하면서 게터가 컬렉션 자체를 반환하도록 한다면, 그 컬렉션을 감싼 클래스가 눈치채지 못하는 상태에서 컬렉션의 원소들이 바뀌어버릴 수 있다.
- 관련된 기능들을 잘게 쪼개고, 묶고, 이동하여 응집도 높게 모듈화를 하는 것이 결국엔 가독성을 높일 수 있다.
- 사이드 이펙트를 줄이기 위해 값을 불변으로 다루는 것이 중요하다.

### 인상 깊은 구절
> "조금씩 변경하고 매번 테스트하는 것은 리팩터링 절차의 핵심이다."

나의 생각: 리팩터링 전 제대로 되었는지 확인을 위해 테스트 코드가 바탕이 되어야 한다!

### 비평적 관점
**좋았던 점**
- 리팩터링의 방법에 무엇들이 있는지 알게 되었다.
- 편함을 위해 사용하던 잘못된 코딩 습관에 대해 깨닫게 되었다.

**아쉬웠던 점**
- js보단 ts로 책을 써줬으면 더 좋았을 것 같다.
- 책의 특성 상 예시 코드를 따라가며 읽기가 힘들었는데 전/후를 github 상에서 PR로 올려서 diff를 보게해주면 좋았을 것 같다.

## ⭐ 종합적 평가
### 책에 대한 평가
- 전체 평점: [5점 만점] 4.5
- 추천하고 싶은 대상:  리팩터링에 관심이 많은 class에 대해 고민해본 개발자
- 이유: 객체 지향과 class에 대한 이야기가 많이 나온다.

**Comment by blan19** [comment]
## 📑 서평
### 한 줄 요약
리팩토링도 코드를 개선하기 위한 체계화된 패턴의 집합이다.

### 주요 내용 정리
[책의 핵심 내용과 인상 깊었던 부분을 정리해주세요]

### 인상 깊은 구절
> p.143 나는 적은 수의 테스트만으로 큰 효과를 얻고 있다. 잘못될까봐 가장 걱정하는 영역을 집중적으로 테스트하는데, 이렇게 해서 테스트에 쏟는 노력의 효과를 극대화하는 것이다.

나의 생각: 매번 테스트 코드를 작성할 때, 귀찮다는 생각이 먼저 들었습니다. 돌이켜 보니 테스트 코드를 효율적으로 작성하지 않고 양으로 때우고 있는 부분이 많아 테스트 코드에 쏟는 노력이 과도했었던 것 같네요
효율적으로 테스트 코드를 작성하는 노력을 기울여야 하겠다는 생각이 들었습니다.

### 비평적 관점
**좋았던 점**
- 리팩토링 방법을 패턴화하여 개선이 필요한 상황에 어떠한 알맞은 리팩토링 기법을 사용해야할지 가이드가 되어 주었습니다.
- 전체적으로 리팩토링 만큼 테스트 중요성의 비중을 많이 두어 평소 비교적 관심이 적었던 테스트까지 생각해보게 해주었습니다.

**아쉬웠던 점**
- 리팩토링이란게 테스트를 강조할 수밖에 없는 구조지만, 리팩터링의 본질보다 테스트에 지나치게 집중하다보니 리팩토링 기법에 집중하기가 조금 힘든 부분이 더러 존재했습니다.

## ⭐ 종합적 평가
### 책에 대한 평가
- 전체 평점: [5점 만점]
- 추천하고 싶은 대상: 리팩토링을 뇌 속에 패턴화하여 자연스럽게 적용하고 싶은 개발자

**Comment by choisohyun** [comment]
## 📑 서평
### 한 줄 요약

영원히 유지되는 코드는 없고 그 때 만든 코드는 그 때의 최선이었다. 냄새 나면 당장 갈아라

### 주요 내용 정리

- 리팩터링 후에 항상 테스트하는 습관을 들여라.
- 테스트를 너무 많이 만들다 보면 오히려 필요한 테스트를 놓치기 쉽다.
- 좋은 소프트웨어 설계의 핵심은 모듈화가 얼마나 잘 되어 있느냐를 뜻하는 모듈성이다. 모듈성이란 프로그램의 어딘가를 수정하려 할 때 해당 기능과 깊이 관련된 작은 일부만 이해해도 가능하게 해주는 능력이다.
- 하나의 값이 여러 목적으로 사용된다면 혼란과 버그를 낳는다.

### 인상 깊은 구절

> 리팩터링은 개발 기간을 단축하고자 하는 것이다.

나의 생각: 리팩터링도 프로덕을 잘 만들기 위한 수단일 뿐 우선순위를 지켜야 한다는 걸 자꾸 까먹게 되는 것 같고 이 프로덕을 관리하는 관점을 달리 보도록 노력해야겠습니다.

### 비평적 관점

**좋았던 점**

- 이 책에 등장하는 리팩터링 기법을 통해 그동안 '리팩토링'으로 뭉쳐서 불리던 작업의 의미를 구체적으로 알 수 있어 좋았습니다.
- 평소 선호하던 리팩토링에서 새로운 관점을 얻을 수 있었습니다. 지금 코드가 악취가 난다고 생각했지만 내 생각보다 괜찮은 코드였던 느낌?

**아쉬웠던 점**

- 예제 언어가 자바스크립트지만 클래스 객체 위주로 설명되어 있었습니다. 함수형이나 프론트엔드 관점에서 리팩토링하는 예제가 있었다면 더 좋았을 것 같습니다.
- `매개변수를 질의 함수로 변경하기, 질의 함수를 매개 함수로 변경하기` 등 역으로 리팩토링하는 경우들이 있었는데, 예제가 있지만 활용하려고 할 때 판단하기가 어렵겠다는 생각이 들었습니다. 책이 주는 한계점 같기도 합니다.(강의로 하나의 큰 객체를 리팩토링했다면 좀 더 낫지 않았을까.. 하는 느낌)

## ⭐ 종합적 평가
### 책에 대한 평가

- 전체 평점: ⭐⭐⭐⭐
- 추천하고 싶은 대상: 리팩토링을 해본 적이 있는 클래스를 주로 쓰는 개발자
- 이유: 사실 책의 초판이 나왔을 시점에 샀었고 보려고 시도한 적이 있었는데, 리팩토링 기법이 너무 많아 이해하기가 어려웠습니다. 지금은 리팩토링도 해봤기 때문에 그 기반으로 이해했던 것 같습니다. 그리고 클래스 위주로 나오기 때문에 객체지향으로 코드를 주로 짠다면 더욱 도움이 될 것 같습니다.

**Comment by suojae** [comment]
## 📑 서평
### 한 줄 요약
[이 책의 핵심 메시지나 가장 큰 인상을 한 문장으로]
함수는 작게, 테스트는 자주

### 주요 내용 정리
[책의 핵심 내용과 인상 깊었던 부분을 정리해주세요]
- 에러가 났던 부분은 테스트 코드부터 작성하고 시작하자.
- 함수길이가 긴 것도 코드 스멜이다.
- 읽기 쉬운 코드는 명확성이 생명이다. 함수들의 exit point가 예측가능하고 명확해지면 읽기 쉬운코드가 된다.


### 인상 깊은 구절
> "The heart of a good software design is its modularity—which is my ability to make most modifications to a program while only having to understand a small part of it."

나의 생각: 사람의 기억력은 한계가 있다. 예를들어 특정 함수를 호출할 때 해당 함수 내용이 많고 또 그함수가 호출하는 다른 함수마저 내용이 많다면 먼저 호출한 함수에 대한 복잡성이 급격하게 올라가서 이해하기 어려운 코드가 될 것이다. 따라서 소프트웨어 디자인의 핵심은 역할과 책임에 맞는 단단한 모듈화이다. 알맞은 역할 분배가 된 작은 모듈들로 구성되어있을 때 비로소 이해하기 쉬운 코드가 된다.

<br/>

>"This is why I often favor objects over records for mutable data. With objects, I can hide what is stored and provide methods for all three values. The user of the object doesn’t need to know or care which is stored and which is calculated."

나의 생각: 가변 데이터를 다룰 때 객체(object)를 사용해야한다. 변화하는 데이터를 객체로 포장하고(캡슐화)  get, set 처리를 해줌으로써 변화하는 데이터를 불변데이터로 바꿔줄 수있고 이는 예측가능성을 높여주는 코드가 된다.


### 비평적 관점
**좋았던 점**
- 정말 다양한 리팩토링 기술 및 노하우가 친절하게 설명되어 있었다. 마치 시니어 개발자가 옆에서 이럴 땐 이렇게 리팩토링하는거란다~ 라고 설명해주는 것 같았다.
- 다양한 기법이 있었던 만큼 코드를 읽는 관점이 확장될 수 있었다. 이런 경우에는 이렇게 짤 수도 있겠구나 아 이경우에는 이렇게도 할 수 있겠구나라는 생각이 확장되었음을 느낄 수 있었다. 
- 테스트 코드 작성의 중요성을 책읽는 내내 되새김질 할 수 있었다.

**아쉬웠던 점**
- 책을 읽으면서 이렇게까지..? 싶은 부분도 있었다. 특히 함수를 잘게 쪼개라는 파트에서는 이렇게까지 세분해서 컴포지셔닝을 할 필요가 있을까? 라는 의문도 있었다. 현실에 적용하기 너무 이상적인 이야기도 있었던 것 같다.

## ⭐ 종합적 평가
### 책에 대한 평가
- 전체 평점: ⭐⭐⭐⭐ [5점 만점] 
- 추천하고 싶은 대상: 코드 퀄리티 향상에 시간을 쏟고싶어하는 개발자
- 이유: 책을 읽는 것도 일종의 트레이드 오프라고 생각한다. 리팩토링 책을 읽을시간에 다른 주제의 기술서적을 읽을 수 있기 때문이다. 이 책은 유지보수가 중요한 환경에서 일하는 개발자, 혹은 이러한 환경에서 일하고 싶은 개발자에게 충분히 시간을 낼만한 가치가 있다고 생각한다. 유지보수성을 높이는 코드란 무엇인가에 대해 관점을 크게 넓힐 수 있어 유지보수가 중요한 코드에 실질적인 변화를 줄 수 있다는 생각이 들었기 때문이다.

**Comment by hamo-o** [comment]
## 📑 서평
### 한 줄 요약
리팩터링은 단순히 코드를 멋지게 보이려고 하는 것이 아니라, 개발 효율성을 높이기 위한 과정이다.

### 주요 내용 정리
- 리팩터링이란, 소프트웨어의 **겉보기 동작은 그대로 유지**한 채, 코드를 이해하고 수정하기 쉽도록 내부 구조를 변경하는 기법
- 기능 추가 모자와 리팩터링 모자를 번갈아 쓰면서 개발하자.
- ”스트라이크 세 번이면 리팩터링하라(삼진 리팩터링)”
- 빨간 막대일 때는 리팩터링하지 말라
- 분리할 부분을 제거했을 때 다른 필드나 메서드들에 논리적인 문제가 없다면 분리할 수 있다는 뜻이다.
- 기능 확장 시 확장해야 할 기능에 따라 서브클래스를 만드는 방식이 달라진다면 클래스를 나눠야 한다는 신호이다.
- 좋은 소프트웨어 설계의 핵심은 모듈성이다.
- 프로그램의 진짜 힘은 데이터 구조이다.

### 인상 깊은 구절
> 무언가 수정하려 할 때는 먼저 수정하기 쉽게 정돈하고 그런 다음 쉽게 수정하자. _켄트 벡

나의 생각: 리팩터링은 "부가적인 일"이라고만 생각했던 기존의 관점을 깨부수는 구절이었다! 무언가 수정하거나 덧붙이기 전에 리팩터링을 하는 습관을 가지려고 노력하기로 했다.

### 비평적 관점
**좋았던 점**
- 책을 한 번 읽었다고 끝이 아니라, 이후 리팩터링의 필요성이 느껴지면 사전식으로 다시 찾아볼 수 있어 좋다.
- 각 기법의 맨 앞 부분에 간단한 그림과 코드 조각을 예시로 줘서 한눈에 리팩터링을 이해하기 편했다.

**아쉬웠던 점**
- 클래스형 위주의 리팩터링 스킬들이 많아서, 실무에서 조금 더 잘 활용할 수 있는 함수형 예제들도 포함되었다면 좋았을 것 같다.

## ⭐ 종합적 평가
### 책에 대한 평가
- 전체 평점: 4/5 (위 아쉬웠던 점을 이유로..)
- 추천하고 싶은 대상: 자바스크립트를 자주 사용하는 개발자 누구나
- 이유: 2장의 리팩터링에 관한 인사이트는 누구나 도움될만한 내용이라고 생각하고, 뒷장의 조금 더 구체적인 가이드들은 필요한 부분만 취사선택하여 읽을 수 있기 때문이다.

**Comment by hyungwookchoi** [comment]
## 📑 서평
### 한 줄 요약
작은 단계를 거쳐 점진적으로 개선하자.

### 주요 내용 정리
- 

### 인상 깊은 구절
> '작은 단계'라 함은 각 단계에서 잘못될 일이 적어졌고, 그래서 할 일도 줄었다는 뜻이다. p.335

나의 생각: 작성하는 코드도, 그 코드를 리팩토링 하는 것도 단계를 쪼개서 '작은 단계'로 변경하는 것이 위험도를 줄이면서도 빠른 수정을 가능하게 한다.

### 비평적 관점
**좋았던 점**
- 좋지 않은 예시를 먼저 보여주고, 여러 단계를 거쳐 좋은 예시로 바꿔가는 과정을 순차적으로 따라갈 수 있어 좋았습니다.
- 테스트 코드의 중요성을 다시 일깨워주고, 꼭 한번에 시작 할 필요 없이 작게라도 시도해봄을 강조해주는 게 좋았습니다.

**아쉬웠던 점**
- 종이책의 한계.. 일지도 모르겠지만, 수정 전/후 코드를 한눈에 파악하기가 힘들었습니다.

## ⭐ 종합적 평가
### 책에 대한 평가
- 전체 평점: ⭐⭐⭐⭐
- 추천하고 싶은 대상: 코드의 지속적인 개선에 관심이 있는 주니어 이상의 개발자
- 이유: 개발에 처음 입문 혹은 신입의 입장에서는 JS에서 많이 다뤄보지 못했을 클래스 문법이나 객체지향에 대해 이해하기가 어려울 것 같고, 리팩토링을 왜, 어떻게 해야 하는지에 대한 고민이 해본 적이 없는 상태에서는 그 의미가 크게 와닿지 않을 것 같습니다. 어느 정도 개발에 익숙하고, 프로덕션 환경에서 유지보수 단계에 접어든 코드를 관리하고 있는 주니어 이상의 개발자에게 더 적합한 책이라고 생각합니다.
